examples/apps/src/ble_bas_peripheral.rs

```diff
@@ -63,80 +66,125 @@ where
-    server.run().await
-}
-
-async fn advertise_task<C: Controller>(
-    mut peripheral: Peripheral<'_, C>,
+/// Stream Events until the connection closes.
+async fn gatt_task<C: Controller>(
     server: &Server<'_, '_, C>,
+    conn: &Connection<'_>,
 ) -> Result<(), BleHostError<C::Error>> {
-    let mut adv_data = [0; 31];
+    let level = server.battery_service.level;
+    loop {
+        if let Either::First(event) = select(conn.next(), server.run()).await {
+            match event {
+                ConnectionEvent::Disconnected { reason } => {
+                    info!("[adv] disconnected: {:?}", reason);
+                    break;
+                }
+                ConnectionEvent::Gatt { event, .. } => match event {
+                    GattEvent::Read { value_handle } => {
+                        if value_handle == level.handle {
+                            let value = server.get(&level);
+                            info!("[gatt] Read Event to Level Characteristic: {:?}", value);
+                        }
+                    }
+                    GattEvent::Write { value_handle } => {
+                        if value_handle == level.handle {
+                            let value = server.get(&level);
+                            info!("[gatt] Write Event to Level Characteristic: {:?}", value);
+                        }
+                    }
+                },
+            }
+        }
+    }
+    info!("[gatt] task finished");
+    Ok(())
+}
+
+/// Create an advertiser to use to connect to a BLE Central, and wait for it to connect.
+async fn advertise<'a, C: Controller>(
+    name: &'a str,
+    peripheral: &mut Peripheral<'a, C>,
+) -> Result<Connection<'a>, BleHostError<C::Error>> {
+    let name = if name.len() > 22 {
+        let truncated_name = &name[..22];
+        info!("Name truncated to {}", truncated_name);
+        truncated_name
+    } else {
+        name
+    };
+    let mut advertiser_data = [0; 31];
     AdStructure::encode_slice(
         &[
             AdStructure::Flags(LE_GENERAL_DISCOVERABLE | BR_EDR_NOT_SUPPORTED),
             AdStructure::ServiceUuids16(&[Uuid::Uuid16([0x0f, 0x18])]),
-            AdStructure::CompleteLocalName(b"Trouble"),
+            AdStructure::CompleteLocalName(name.as_bytes()),
         ],
-        &mut adv_data[..],
+        &mut advertiser_data[..],
     )?;
+    let mut advertiser = peripheral
+        .advertise(
+            &Default::default(),
+            Advertisement::ConnectableScannableUndirected {
+                adv_data: &advertiser_data[..],
+                scan_data: &[],
+            },
+        )
+        .await?;
+    info!("[adv] advertising");
+    let conn = advertiser.accept().await?;
+    info!("[adv] connection established");
+    Ok(conn)
+}
+
+/// Example task to use the BLE notifier interface.
+async fn example_application_task<C: Controller>(server: &Server<'_, '_, C>, conn: &Connection<'_>) {
```

Comment:
I think this is the task that should do all processing of the connection, maybe also rename it connection_task or something?

Reply:
I wanted to separate out the connection from this example task, as you could have several tasks all associated with the same connection. (i.e. one for each sensor on your system such as the example below from my Gamepad demo)

```rust
    // Main loop
    loop {
        display.display(QuestionMark, Duration::from_secs(2)).await;
        // advertise for connections
        if let Ok(conn) = advertiser.advertise().await {
            let pause = Duration::from_secs(1);
            speaker.play_tune(Tune::Connect).await;
            display.display_blocking(Heart, pause).await;


            let gatt = gatt_server_task(server, &conn);
            let buttons = buttons_task(&mut gamepad_buttons, &conn, &display);
            let analog = analog_stick_task(server, &conn, &mut analog_stick, &display);
            embassy_futures::select::select3(gatt, buttons, analog).await;
            speaker.play_tune(Tune::Disconnect).await;
        }
```